### ISSUE TITLE
Improve load balancing on multiple GPUs

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#3457]
This is a followup to #3333.  We made changes there that significantly improved multi-GPU performance.  A major limiting factor at this point seems to be load balancing.  Ideally we want all GPUs to finish their work at the same time, but in practice we usually get a lot of variation so some GPUs spend extra time sitting idle.  We should investigate whether we can improve that.

**Comment by peastman** [comment]
In order to improve the load balancing, I need to accurately measure how long it takes for each GPU to complete its computations and data transfers.  Currently that's done by calling `cuCtxSynchronize()` and measuring the time on the host.  It works, but not all that well.  The synchronization makes it expensive, so I don't want to do it too frequently.  And it isn't a very accurate way of measuring when the GPU has finished its work.

An alternative is `cuEventElapsedTime()`.  In principle that should allow direct timing measurements from the GPU without involving the host.  But according to the [documentation](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EVENT.html#group__CUDA__EVENT_1gdfb1178807353bbcaa9e245da497cf97) there are two major problems with it.  First, the time resolution is only about 0.5 ms, which isn't very good for this purpose.  And second, if you use any stream other than the default one its result may be wrong.  I think that mostly rules it out.

Is there a better way of measuring time?  The CUDA profiler reports exact times when each operation begins and ends with sub-microsecond resolution.  That suggests this information must be available somewhere.  Is there a way to retrieve it?

cc @dmclark17

**Comment by dmclark17** [comment]
The time resolution of `cuEventElapsedTime` should be around 0.5 microseconds. The non-NULL stream behavior does seem potentially problematic. I will look into other solutions!

**Comment by peastman** [comment]
@dmclark17 did you ever figure out if there's a way to get accurate timings that works with multiple streams?

**Comment by dmclark17** [comment]
Sorry, I lost track of this—I think we might be able to use CUDA events for this use-case. It looks like CUPTI could also be a possibility, but I am not very familiar with the API: [https://docs.nvidia.com/cupti/r_overview.html#r_overview](https://docs.nvidia.com/cupti/r_overview.html#r_overview)

This GTC talk goes into more detail about the potential issues with CUDA event timing: https://developer.nvidia.com/gtc/2019/video/s9956 starting around 25:35 or slide 44. Most of the issues have to do with multiple streams leading to other kernel launches being included in the timing. I don't think the PME stream will interfere since its work will be completely contained between the two events on the leader GPU. I'm not sure if the same caveats apply to streams on other devices—I'll look more into this.

**Comment by dmclark17** [comment]
I think CUDA events will produce reasonable accurate results. While there are technically multiple streams per device, I don't think they will be an issue given what we would like to time. I believe the CUPTI API would give more accurate results but would likely be more work. I can try to prototype the event based approach if you'd like.

**Comment by peastman** [comment]
I tried using events a while back, but I couldn't get any kind of meaningful numbers out of them.  I'll take another look and see if I can figure out what was wrong.

**Comment by dmclark17** [comment]
I think it needs a `cuStreamSynchronize` call near line 266—I believe it may be calling `cuEventElapsedTime` before one of the events has been completed

**Comment by peastman** [comment]
Just above at line 247 it waits for all the events:

```c++
for (int i = 0; i < data.contexts.size(); i++)
    cuStreamWaitEvent(cu.getCurrentStream(), peerCopyEvent[i], 0);
```

Shouldn't that be sufficient?

**Comment by dmclark17** [comment]
I believe those are non-blocking on the host side—it is making the stream wait on the given event

**Comment by peastman** [comment]
Should I first call `cuStreamWaitEvent()` and then `cuStreamSynchronize()` for each stream?  Or just the latter?

**Comment by dmclark17** [comment]
Because there are the `cuStreamWaitEvent` calls, I think we only need to synchronize on `cu.getCurrentStream()`. The current stream will wait for the events to complete, so, if we synchronize on the current stream, the host will wait for the events to complete.

**Comment by dmclark17** [comment]
Based on comparing the reported times with an Nsight System profile, it seemed like the PME stream wasn't being included for the 0th device. Maybe `peerCopyEvent[0]` was being recorded before `pmeSyncEvent`? 

Rerecording the `peerCopyEvent[0]` in `CudaParallelCalcForcesAndEnergyKernel::finishComputation` seemed to yield better load balancing. I.e. changing lines 246-248 to be something like this:

```
    if (cu.getPlatformData().peerAccessSupported) {
        cuEventRecord(peerCopyEvent[0], cu.getCurrentStream());
        for (int i = 0; i < data.contexts.size(); i++)
            cuStreamWaitEvent(cu.getCurrentStream(), peerCopyEvent[i], 0);
    }
```

Not sure if there is a better way to include the PME stream

**Comment by peastman** [comment]
It looks to me like the PME stream ought to be included.  We record an event when all its work is finished:

https://github.com/openmm/openmm/blob/c3c8ec55e87f515e37151b346a7535e5e9c6b99b/platforms/cuda/src/CudaKernels.cpp#L1368-L1371

When `finishComputation()` is called on the `CudaParallelCalcForcesAndEnergyKernel`, the first thing it does is submit tasks to be run in parallel to finish each of the separate device computations.  Then it calls `syncContexts()` to wait until those are done.

https://github.com/openmm/openmm/blob/c3c8ec55e87f515e37151b346a7535e5e9c6b99b/platforms/cuda/src/CudaParallelKernels.cpp#L238-L244

Each one of those tasks will execute the NonbondedForce's post-computation, which waits on the event.

https://github.com/openmm/openmm/blob/c3c8ec55e87f515e37151b346a7535e5e9c6b99b/platforms/cuda/src/CudaKernels.cpp#L533

So by the time `syncContexts()` returns, `cuStreamWaitEvent()` should have been called for the PME stream.

**Comment by peastman** [comment]
Here's what I'm seeing when running on four GPUs.

If I add the call to `cuStreamSynchronize()` you suggested, the benchmark immediately exits with an error that the particle coordinates have become nan.  Something is clearly messed up there.

If I add the call, and also re-record the event as you suggested, then the error goes away.  The load balancing looks a bit more plausible in this case.  It moves all work off GPU 0 (which could potentially make sense, since that GPU is having to run the reciprocal space calculation) and distributes it unevenly between the other three: about one quarter each on GPUs 1 and 2, and half the work on GPU 3.  However, the overall performance is slightly slower than with the original version of the code, so clearly it is doing a less good job of load balancing.  Looking at the timings it registers, GPU 3 is always, without exception, reported as being the first one to finish.

**Comment by dmclark17** [comment]
We are computing the time based on the `peerCopyEvent[0]` event which is recorded within the `CudaParallelCalcForcesAndEnergyKernel::FinishComputationTask` function and on the `peerCopyStream[0]` stream. That would happen on a different stream than the `syncContexts`, right?

Interesting—I was only testing with 2 GPUs, but I don't think I saw any crashes when just adding the `cuStreamSynchronize` call. I will try on 4 GPUs as well!

**Comment by peastman** [comment]
The time for device `i` is computed between `event`, which is recorded in `CudaParallelCalcForcesAndEnergyKernel::beginComputation()`, and `peerCopyEvent[i]`.

`syncContexts()` is a host-side synchronization.  Every CudaContext has a worker thread that can be used to asynchronously do work.  `syncContexts()` waits until all those threads have finished all their work.  In this case, that includes a `CudaCalcNonbondedForceKernel::SyncStreamPostComputation` which (among other things) calls `cuStreamWaitEvent()`.

**Comment by dmclark17** [comment]
I got sidetracked with some other work—I will look into this more in the coming days

**Comment by dmclark17** [comment]
I still need to think more about the synchronization; however, with both the cuStreamSynchronize and event re-record, I saw a similar behavior to you for 4 GPUs. On a DGX V100 node and cellulose, it assigned 5% of the work to GPU0, 20% to GPU1, 20% to GPU2, and 60% to GPU3. 

Here is a screenshot of the nsight system timeline for that run:

![cellulose_dgx](https://user-images.githubusercontent.com/30421289/175344594-ada1a5da-432c-47be-88ba-409eb958b149.png)

For timesteps where `findBLocksWithInteractions` does work, the load balancing seems pretty good; however, GPU0 seems to be overloaded for other timesteps. 

The node topology looks like this:

```
        GPU0    GPU1    GPU2    GPU3   
GPU0     X      NV1     NV1     NV2    
GPU1    NV1      X      NV2     NV1     
GPU2    NV1     NV2      X      NV2     
GPU3    NV2     NV1     NV2      X     
```

The data transfers from GPU0->GPU1 and GPU0->GPU2 share an NVLink connection, so they achieve about half the bandwidth of the GPU0->GPU3 transfers—I confirmed this in the profiler. This explains why GPU3 is receiving a larger fraction of the work.

**Comment by peastman** [comment]
Any idea what could be causing it to serialize them?  Clearly the hardware is capable of overlapping memcpys, since it overlaps the ones for forces.  The profile shows each one is on a different stream, so that shouldn't be the problem.

**Comment by dmclark17** [comment]
I've been looking into this a little more, and this is slightly orthogonal to the measurement question. Would it be possible to have different work fractions for timesteps where `findBlocksWithInteractions` does work? In the screenshot below, the load balancing seems pretty good for expensive `findBlocksWithInteractions` timesteps; however, when the neighbor list is reused, the "offload/nonbonded" GPUs are left idle. 

Also, we might run into the case where the PME calculation is on the critical path even if no nonbonded work is assigned to the "PME/main" GPU. Increasing the pme-cutoff should make the PME calculation faster, right? Or–will it not make a meaningful difference?

![Screenshot 2022-08-22 161212](https://user-images.githubusercontent.com/30421289/186036221-b6899a01-f146-489d-a051-e7d87f68f640.png)

**Comment by peastman** [comment]
That would be challenging.  The division of work between GPUs is implemented as part of building the neighbor list (the `startBlockIndex` and `numBlocks` arguments to `findBlocksWithInteractions()`).  To have two different divisions, we would need to build two different neighbor lists, which would be expensive.  It would need to build one neighbor list to use on that step, and a separate one to use on following steps in which the list doesn't get rebuilt.

> Increasing the pme-cutoff should make the PME calculation faster, right?

Correct.  It will reduce the size of the grid, which will make the FFTs faster, and also make charge spreading slightly faster due to better cache performance.

**Comment by dmclark17** [comment]
Makes sense—probably best to avoid constructing two neighbor lists. Looking at the profiles, the neighbor list is regenerated every other time step for cellulose and STMV, so every other timestep has suboptimal load balancing. If we tuned the load balancing for the non-neighbor list timesteps and increased the number of timesteps for which the neighbor lists are reused, maybe we could on average have better load balancing.

I will investigate this as well as tuning the pme-cutoff—unclear to me if either will payoff since they'll probably adversely impact single GPU performance.

**Comment by peastman** [comment]
The load balancing should reflect average performance over all steps, both ones with and without neighbor list rebuilds.  If something changes that frequency, I think it will automatically get taken into account.

**Comment by dmclark17** [comment]
During the load balancing phase, the neighbor list will be regenerated after every timestep, right? Since the number of atoms assigned to each device changes, the neighbor list will need to be computed. This will tune the percentages for neighbor list rebuild steps I think.

**Comment by peastman** [comment]
Good point.  Maybe we should only measure and load balance on a random subset of steps?  That will give a more representative sample.

**Comment by dmclark17** [comment]
That would be interesting—I can look into it! 

Interestingly, I am finding that the performance characteristics of STMV has changed since these benchmarks https://github.com/openmm/openmm/pull/2989#issuecomment-768790704 were taken. With the current 8% pad, I am measuring 15.8268 ns/day; however, with a pad of 15% I am measuring 17.7608 ns/day on a Titan V. There is a similar improvement on A100 too

**Comment by peastman** [comment]
I made some improvements to load balancing in #4013.  The effect isn't huge, but it does help.
